tika/server.go

```diff
@@ -0,0 +1,230 @@
+
+// cmder is used to stub out *exec.Cmd for testing.
+var cmder commander = exec.CommandContext
+
+// Start starts the given server. Start will start a new Java process. The
+// caller must call cancel() to shut down the process when finished with the
+// Server. The given Context is used for the Java process, not for cancellation
+// of startup.
+func (s *Server) Start(ctx context.Context) (cancel func(), err error) {
+	ctx, cancel = context.WithCancel(ctx)
+	cmd := cmder(ctx, "java", "-jar", s.jar, "-p", s.port)
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if err := s.waitForStart(ctx); err != nil {
+		cancel()
+		buf, readErr := ioutil.ReadAll(stderr)
+		if readErr != nil {
+			return nil, fmt.Errorf("error reading stderr: %v", readErr)
+		}
+		// Report stderr since sometimes the server says why it failed to start.
+		return nil, fmt.Errorf("error starting server: %v\nserver stderr:\n\n%v", err, string(buf))
+	}
+	return cancel, nil
+}
+
+// waitForServer waits until the given Server is responding to requests.
+// waitForStart returns an error if the server does not respond within the
+// timeout set by WithStartupTimeout or if ctx is Done() first.
+func (s Server) waitForStart(ctx context.Context) error {
+	c := NewClient(nil, s.url)
+	ctx, cancel := context.WithTimeout(ctx, s.startupTimeout)
+	defer cancel()
+	for {
+		select {
+		case <-time.Tick(500 * time.Millisecond):
+			if _, err := c.Version(ctx); err == nil {
+				return nil
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func validateFileMD5(path, wantH string) (bool, string) {
+	f, err := os.Open(path)
+	if err != nil {
+		return false, ""
+	}
+	defer f.Close()
+
+	h := md5.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return false, ""
+	}
+	md5 := fmt.Sprintf("%x", h.Sum(nil))
+	return md5 == wantH, md5
+}
+
+// A Version represents a Tika Server version.
+type Version string
+
+// Supported versions of Tika Server.
+const (
+	Version114 Version = "1.14"
+	Version115 Version = "1.15"
+	Version116 Version = "1.16"
+)
+
+var md5s = map[Version]string{
+	Version114: "39055fc71358d774b9da066f80b1141c",
+	Version115: "80bd3f00f05326d5190466de27d593dd",
+	Version116: "6a549ce6ef6e186e019766059fd82fb2",
+}
+
+// DownloadServer downloads and validates the given server version,
+// saving it at path. DownloadServer returns an error if it could
+// not be downloaded/validated. Valid values for the version are 1.14.
+// It is the caller's responsibility to remove the file when no longer needed.
+// If the file already exists and has the correct MD5, DownloadServer will
+// do nothing.
+func DownloadServer(ctx context.Context, version Version, path string) error {
+	wantH := md5s[version]
+	if wantH == "" {
+		return fmt.Errorf("unsupported Tika version: %s", version)
+	}
+
+	if _, err := os.Stat(path); err == nil {
```

Comment:
just let `validateFileMD5` fail? checking for existence before opening is pointless, the file could be deleted in between

Reply:
Good catch!